File: src/whirlwind/graph/dijkstra.cpp
#include <cstdint>

#include <nanobind/nanobind.h>

#include <whirlwind/common/type_traits.hpp>
#include <whirlwind/graph/csr_graph.hpp>
#include <whirlwind/graph/dijkstra.hpp>
#include <whirlwind/graph/rectangular_grid_graph.hpp>

#include "iterable.hpp"

namespace whirlwind::bindings {

namespace nb = nanobind;
using namespace nb::literals;

template<class Class, class... Extra>
void
dijkstra_attrs_and_methods(nb::class_<Class, Extra...>& cls)
{
    using Graph = typename Class::graph_type;

    // Constructors.
    cls.def(nb::init<const Graph&>(), "graph"_a);

    // Methods.
    cls.def("has_reached_vertex", &Class::has_reached_vertex, "vertex"_a);
    cls.def("has_visited_vertex", &Class::has_visited_vertex, "vertex"_a);
    cls.def("label_vertex_reached", &Class::label_vertex_reached, "vertex"_a);
    cls.def("label_vertex_visited", &Class::label_vertex_visited, "vertex"_a);
    cls.def("reached_vertices", &Class::reached_vertices);
    cls.def("visited_vertices", &Class::reached_vertices);
    cls.def("distance_to_vertex", &Class::distance_to_vertex, "vertex"_a);
    cls.def("set_distance_to_vertex", &Class::set_distance_to_vertex, "vertex"_a,
            "distance"_a);
    cls.def("add_source", &Class::add_source, "source"_a);
    cls.def("pop_next_unvisited_vertex", &Class::pop_next_unvisited_vertex);
    cls.def("visit_vertex", &Class::visit_vertex, "vertex"_a, "distance"_a);
    cls.def("relax_edge", &Class::relax_edge, "edge"_a, "tail"_a, "head"_a,
            "distance"_a);
    cls.def("done", &Class::done);
    cls.def("reset", &Class::reset);
}

template<class Distance, class Graph>
void
dijkstra_inst(nb::module_& m, const char* name)
{
    using Class = Dijkstra<Distance, Graph>;
    using Parent = Class::forest_type;
    using ReachedVertices =
            remove_cvref_t<decltype(std::declval<Class>().reached_vertices())>;
    using VisitedVertices =
            remove_cvref_t<decltype(std::declval<Class>().visited_vertices())>;

    auto dijkstra = nb::class_<Class, Parent>(m, name);
    dijkstra_attrs_and_methods(dijkstra);

    if (!nb::type<ReachedVertices>().is_valid()) {
        auto reached_vertices =
                nb::class_<ReachedVertices>(dijkstra, "_ReachedVertices");
        common_iterable_attrs_and_methods(reached_vertices);
    }

    if (!nb::type<VisitedVertices>().is_valid()) {
        auto visited_vertices =
                nb::class_<VisitedVertices>(dijkstra, "_VisitedVertices");
        common_iterable_attrs_and_methods(visited_vertices);
    }
}

void
dijkstra(nb::module_& m)
{
    using f64 = double;
    using i64 = std::int64_t;

    dijkstra_inst<f64, CSRGraph<>>(m, "Dijkstra__f64_CSRGraph");
    dijkstra_inst<i64, CSRGraph<>>(m, "Dijkstra__i64_CSRGraph");
    dijkstra_inst<f64, RectangularGridGraph<>>(m, "Dijkstra__f64_RectangularGridGraph");
    dijkstra_inst<i64, RectangularGridGraph<>>(m, "Dijkstra__i64_RectangularGridGraph");
}

} // namespace whirlwind::bindings

File: src/whirlwind/graph/_dijkstra.py
import enum
from typing import Generic, TypeVar
from collections.abc import Iterable

from . import _lib
from ._forest import Forest

__all__ = [
    "Dijkstra",
    "DistanceType",
]


# FIXME
Graph = TypeVar("Graph")
Vertex = TypeVar("Vertex")
Edge = TypeVar("Edge")
Distance = TypeVar("Distance")


class DistanceType(enum.Enum):
    INT = enum.auto()
    REAL = enum.auto()


def _make_dijkstra_impl(graph, distance_type):
    if distance_type == DistanceType.REAL:
        return _lib.Dijkstra(graph._impl, float())
    if distance_type == DistanceType.INT:
        return _lib.Dijkstra(graph._impl, int())
    raise ValueError


class Dijkstra(Forest, Generic[Graph, Distance]):

    def __init__(self, graph: Graph, distance_type: DistanceType = DistanceType.REAL):
        self._impl = _make_dijkstra_impl(graph, distance_type)
        self._distance_type = distance_type

    @property
    def distance_type(self) -> DistanceType:
        return self._distance_type

    def has_reached_vertex(self, vertex: Vertex) -> bool:
        return self._impl.has_reached_vertex(vertex)

    def has_visited_vertex(self, vertex: Vertex) -> bool:
        return self._impl.has_visited_vertex(vertex)

    def label_vertex_reached(self, vertex: Vertex) -> None:
        self._impl.label_vertex_reached(vertex)

    def label_vertex_visited(self, vertex: Vertex) -> None:
        self._impl.label_vertex_visited(vertex)

    def reached_vertices(self) -> Iterable[Vertex]:
        return self._impl.reached_vertices()

    def visited_vertices(self) -> Iterable[Vertex]:
        return self._impl.visited_vertices()

    def distance_to_vertex(self, vertex: Vertex) -> Distance:
        return self._impl.distance_to_vertex(vertex)

    def set_distance_to_vertex(self, vertex: Vertex, distance: Distance) -> None:
        self._impl.set_distance_to_vertex(vertex, distance)

    def add_source(self, source: Vertex) -> None:
        self._impl.add_source(source)

    def pop_next_unvisited_vertex(self) -> tuple[Vertex, Distance]:
        return self._impl.pop_next_unvisited_vertex()

    def visit_vertex(self, vertex: Vertex, distance: Distance) -> None:
        self._impl.visit_vertex(vertex, distance)

    def relax_edge(self, edge: Edge, tail: Vertex, head: Vertex, distance: Distance) -> None:
        self._impl.relax_edge(edge=edge, tail=tail, head=head, distance=distance)
